File: cpp/test_hmac.cpp
#include "include/hmac.hpp"
#include <iostream>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

// TODO: HMAC salt - In actual implementation, load from environment variables or configuration file
const std::string salt = "my_secret_salt_value_!@#$%^&*"; // hmac key

struct Body {
    int userId;
    std::string username;
    std::string email;
    std::string role;
    int createdAt;
};

void testSha256() {
    Body body = {
        1234,
        "홍길동",
        "user@example.com",
        "admin",
        1234567890
    };

    json bodyJson = {
        {"userId", body.userId},
        {"username", body.username},
        {"email", body.email},
        {"role", body.role},
        {"createdAt", body.createdAt}
    };

    std::string jsonString = bodyJson.dump();
    std::cout << jsonString << std::endl;

    std::string hashString = hmac_utils::HmacService::GenerateHmac(jsonString);
    std::cout << "hashString: " << hashString << std::endl; // expected X-HMAC-Signature: f96cf60394f6b8ad3c6de2d5b2b1d1a540f9529082a8eb9cee405bfbdd9f37a1
}

int main() {
    testSha256();
    return 0;
} 

File: typescript/test_hmac.js
const crypto = require('crypto');

// Base64 URL decoding function (as per guide specification)
function base64UrlDecode(str) {
  // Convert URL safe base64 to standard base64
  str = str.replace(/-/g, '+').replace(/_/g, '/');
  // Add padding (if needed)
  while (str.length % 4) {
    str += '=';
  }
  return Buffer.from(str, 'base64');
}

function hmacSha256(data, salt) {
  return crypto.createHmac('sha256', base64UrlDecode(salt)).update(data).digest('hex');
}

// Define JSON object request body
const requestBody = {
  userId: 1234,
  username: '홍길동',
  email: 'user@example.com',
  role: 'admin',
  createdAt: 1234567890,
};

const salt = 'my_secret_salt_value_!@#$%^&*'; // hmac key
const jsonString = JSON.stringify(requestBody);

console.log('JSON string:', jsonString);
// Output result
console.log('HMAC-SHA256:', hmacSha256(jsonString, salt)); 
// expected X-HMAC-Signature: f96cf60394f6b8ad3c6de2d5b2b1d1a540f9529082a8eb9cee405bfbdd9f37a1 